sklearn/inspection/_permutation_importance.py

```diff
@@ -30,9 +40,31 @@ def _calculate_permutation_scores(
     # (memmap). X.copy() on the other hand is always guaranteed to return a
     # writable data-structure whose columns can be shuffled inplace.
     X_permuted = X.copy()
+    y_mod = y.copy()
+
+    # Draw from rows if not full data is used and subset X and y
+    if max_samples is not X.shape[0]:
+        row_indices = _generate_indices(
+            random_state=random_state,
+            bootstrap=False,
+            n_population=X.shape[0],
+            n_samples=max_samples,
+        )
+        if hasattr(X_permuted, "iloc"):
+            X_permuted = X_permuted.iloc[row_indices]
+            if hasattr(y_mod, "iloc"):
+                y_mod = y_mod.iloc[row_indices]
+            else:
+                y_mod = y_mod[row_indices]
+        else:
+            X_permuted = X_permuted[row_indices]
+            if hasattr(y_mod, "iloc"):
+                y_mod = y_mod.iloc[row_indices]
+            else:
+                y_mod = y_mod[row_indices]
```

Comment:
Thx, I wasn't aware of _safe_indexing

Reply:
I do not understand why we do not need a copy of y since we actually overwritte y. In the original comment is says "Work on a copy of X to to ensure thread-safety in case of threading based". Why does this not apply to y?
Also, if we could overwrite y we could overwrite X as well?

Reply:
The story about `X` is that we do permutation inplace latter on:

```python
X_permuted[:, col_idx] = X_permuted[shuffling_idx, col_idx]
```

With `y`, doing the indexing will just create a new object.

Reply:
Actually, we could do:

```suggestion
    # Draw from rows if not full data is used and subset X and y
    if max_samples < X.shape[0]:
        row_indices = _generate_indices(
            random_state=random_state,
            bootstrap=False,
            n_population=X.shape[0],
            n_samples=max_samples,
        )
        X_permuted = _safe_indexing(X, row_indices, axis=0)
        y = _safe_indexing(y, row_indices, axis=0)
    else:
        X_permuted = X.copy()
```

and remove the original copy in the case that we use `_safe_indexing`. Otherwise, we will possibly twice `X` in memory and `X` subsampled. Now, we would have only `X` and `X` subsampled.

Reply:
Got it 👍